File: src/mnemonic/converter/image.py
"""画像変換モジュール

TLG画像形式のデコードおよび変換機能を提供する。
吉里吉里2エンジンで使用されるTLG5/TLG6形式の画像を
標準的な画像形式（PNG等）に変換する。
また、BMP/JPG/PNG形式からWebP/PNG形式への変換機能も提供する。
"""

from dataclasses import dataclass
from enum import Enum
from pathlib import Path

from PIL import Image

from mnemonic.converter.base import BaseConverter, ConversionResult, ConversionStatus
from mnemonic.converter.tlg import TLG5Decoder, TLG6Decoder


class TLGVersion(Enum):
    """TLG画像のバージョン

    吉里吉里2エンジンで使用されるTLG画像形式のバージョンを表す。
    TLG5とTLG6の2つのバージョンが存在する。
    """

    TLG5 = "TLG5"
    TLG6 = "TLG6"
    UNKNOWN = "UNKNOWN"


class QualityPreset(Enum):
    """WebP変換時の品質プリセット

    画像変換時のWebP品質値を定義する列挙型。
    HIGH/MEDIUM/LOWの3段階の品質レベルを提供する。
    """

    HIGH = 95
    MEDIUM = 85
    LOW = 70


class OutputFormat(Enum):
    """画像出力形式

    ImageConverterの出力形式を定義する列挙型。
    krkrsdl2がWebP未対応のため、PNG出力をデフォルトとする。
    """

    WEBP = "webp"
    PNG = "png"


@dataclass(frozen=True)
class TLGInfo:
    """TLG画像のメタ情報

    TLG画像ファイルから読み取ったメタ情報を保持する不変データクラス。

    Attributes:
        version: TLGバージョン（TLG5またはTLG6）
        width: 画像の幅（ピクセル）
        height: 画像の高さ（ピクセル）
        has_alpha: アルファチャンネルの有無
    """

    version: TLGVersion
    width: int
    height: int
    has_alpha: bool


class TLGImageDecoder:
    """TLG画像デコーダー

    TLG形式の画像ファイルを読み込み、PIL.Imageオブジェクトに変換する。
    TLG5およびTLG6形式に対応。
    """

    # TLGマジックバイト
    TLG5_MAGIC = b"TLG5.0\x00raw\x1a"
    TLG6_MAGIC = b"TLG6.0\x00raw\x1a"

    def __init__(self) -> None:
        """TLGImageDecoderを初期化する"""
        self._tlg5_decoder = TLG5Decoder()
        self._tlg6_decoder = TLG6Decoder()

    def is_tlg_file(self, file_path: Path) -> bool:
        """指定されたファイルがTLG形式かどうかを判定する

        ファイルのマジックバイトを読み取り、TLG5またはTLG6形式かを判定する。

        Args:
            file_path: 判定対象のファイルパス

        Returns:
            TLG形式の場合True、そうでない場合False
        """
        if not file_path.exists():
            return False

        try:
            with open(file_path, "rb") as f:
                header = f.read(len(self.TLG5_MAGIC))
                return header == self.TLG5_MAGIC or header == self.TLG6_MAGIC
        except OSError:
            return False

    def _detect_version(self, data: bytes) -> TLGVersion:
        """TLGデータのバージョンを判別する

        Args:
            data: TLG画像データ

        Returns:
            TLGバージョン
        """
        if data.startswith(self.TLG5_MAGIC):
            return TLGVersion.TLG5
        elif data.startswith(self.TLG6_MAGIC):
            return TLGVersion.TLG6
        else:
            return TLGVersion.UNKNOWN

    def get_info(self, file_path: Path) -> TLGInfo:
        """TLG画像のメタ情報を取得する

        Args:
            file_path: TLG画像ファイルのパス

        Returns:
            TLG画像のメタ情報

        Raises:
            ValueError: TLG形式でないファイルの場合
            FileNotFoundError: ファイルが存在しない場合
        """
        if not file_path.exists():
            raise FileNotFoundError(f"ファイルが見つかりません: {file_path}")

        data = file_path.read_bytes()
        version = self._detect_version(data)

        if version == TLGVersion.TLG5:
            tlg5_header = self._tlg5_decoder.parse_header(data)
            return TLGInfo(
                version=TLGVersion.TLG5,
                width=tlg5_header.width,
                height=tlg5_header.height,
                has_alpha=tlg5_header.colors == 4,
            )
        elif version == TLGVersion.TLG6:
            tlg6_header = self._tlg6_decoder.parse_header(data)
            return TLGInfo(
                version=TLGVersion.TLG6,
                width=tlg6_header.width,
                height=tlg6_header.height,
                has_alpha=tlg6_header.colors == 4,
            )
        else:
            raise ValueError(f"TLG形式ではありません: {file_path}")

    def decode(self, file_path: Path) -> Image.Image:
        """TLG画像をデコードしてPIL.Imageオブジェクトを返す

        Args:
            file_path: TLG画像ファイルのパス

        Returns:
            デコードされたPIL.Imageオブジェクト

        Raises:
            ValueError: TLG形式でないファイルの場合
            FileNotFoundError: ファイルが存在しない場合
            NotImplementedError: TLG6形式の場合（デコーダーが未実装）
        """
        if not file_path.exists():
            raise FileNotFoundError(f"ファイルが見つかりません: {file_path}")

        data = file_path.read_bytes()
        version = self._detect_version(data)

        if version == TLGVersion.TLG5:
            return self._tlg5_decoder.decode(data)
        elif version == TLGVersion.TLG6:
            # TLG6Decoder.decode()はNotImplementedErrorを投げる
            return self._tlg6_decoder.decode(data)
        else:
            raise ValueError(f"TLG形式ではありません: {file_path}")

    def decode_to_file(self, source: Path, dest: Path) -> None:
        """TLG画像をデコードしてファイルに保存する

        Args:
            source: TLG画像ファイルのパス
            dest: 出力先ファイルのパス（拡張子で出力形式を決定）

        Raises:
            ValueError: TLG形式でないファイルの場合
            FileNotFoundError: ファイルが存在しない場合
            NotImplementedError: TLG6形式の場合（デコーダーが未実装）
        """
        image = self.decode(source)
        try:
            dest.parent.mkdir(parents=True, exist_ok=True)
            image.save(dest)
        finally:
            image.close()


class ImageConverter(BaseConverter):
    """画像変換クラス

    BMP/JPG/PNG/TLG形式の画像をWebP/PNG形式に変換する。
    品質設定やアルファチャンネルの取り扱いをカスタマイズ可能。

    Attributes:
        output_format: 出力形式（WebPまたはPNG）
        quality: WebP出力時の品質値（0-100）
        lossless_alpha: アルファチャンネルをロスレスで保存するか
    """

    def __init__(
        self,
        output_format: OutputFormat = OutputFormat.PNG,
        quality: QualityPreset | int = QualityPreset.HIGH,
        lossless_alpha: bool = True,
    ) -> None:
        """ImageConverterを初期化する

        Args:
            output_format: 出力形式（デフォルトはPNG、krkrsdl2互換のため）
            quality: WebP品質（プリセットまたは0-100の整数）
            lossless_alpha: アルファチャンネルをロスレスで保存するか（WebP時のみ使用）
        """
        self._output_format = output_format
        if isinstance(quality, QualityPreset):
            self._quality = quality.value
        else:
            self._quality = quality
        self._lossless_alpha = lossless_alpha
        self._tlg_decoder = TLGImageDecoder()

    @property
    def output_format(self) -> OutputFormat:
        """出力形式を返す"""
        return self._output_format

    @property
    def quality(self) -> int:
        """WebP品質値を返す"""
        return self._quality

    @property
    def lossless_alpha(self) -> bool:
        """ロスレスアルファ設定を返す"""
        return self._lossless_alpha

    @property
    def supported_extensions(self) -> tuple[str, ...]:
        """対応する拡張子のタプルを返す

        Returns:
            対応する拡張子のタプル（.tlgのみ）
            JPEG/PNG/BMPはkrkrsdl2でネイティブサポートのため変換不要
        """
        return (".tlg",)

    def can_convert(self, file_path: Path) -> bool:
        """このConverterで変換可能なファイルかを判定する

        ファイル拡張子に基づいて判定を行う。

        Args:
            file_path: 判定対象のファイルパス

        Returns:
            変換可能な場合True、そうでない場合False
        """
        ext = file_path.suffix.lower()
        return ext in self.supported_extensions

    def convert(self, source: Path, dest: Path) -> ConversionResult:
        """画像ファイルを指定された形式に変換する

        BMP/JPG/PNG/TLG形式の画像をWebP/PNG形式に変換し、
        指定されたパスに出力する。

        Args:
            source: 変換元ファイルのパス
            dest: 変換先ファイルのパス

        Returns:
            変換結果を表すConversionResultオブジェクト
        """
        self._validate_source(source)
        bytes_before = self._get_file_size(source)

        ext = source.suffix.lower()
        # TLGファイルはTLGImageDecoderでデコード
        img = self._tlg_decoder.decode(source) if ext == ".tlg" else Image.open(source)

        try:
            if self._output_format == OutputFormat.PNG:
                result = self._save_as_png(img, dest, source, bytes_before)
            else:
                result = self._save_as_webp(img, dest, source, bytes_before)
        finally:
            img.close()

        return result

    def convert_from_image(self, image: Image.Image, dest: Path) -> ConversionResult:
        """PIL.Imageオブジェクトを指定形式で保存する

        既にメモリ上にある画像オブジェクトを直接WebP/PNG形式で保存する。
        TLGデコード後の画像変換等に使用。

        Args:
            image: 変換元のPIL.Imageオブジェクト
            dest: 変換先ファイルのパス

        Returns:
            変換結果を表すConversionResultオブジェクト
        """
        # PIL.Imageからの変換はsource_pathが無いのでdestをダミーとして使用
        if self._output_format == OutputFormat.PNG:
            return self._save_as_png(image, dest, dest, bytes_before=0)
        else:
            return self._save_as_webp(image, dest, dest, bytes_before=0)

    def _save_as_webp(
        self,
        image: Image.Image,
        dest: Path,
        source: Path,
        bytes_before: int,
    ) -> ConversionResult:
        """画像をWebP形式で保存する内部メソッド

        Args:
            image: 保存するPIL.Imageオブジェクト
            dest: 保存先パス
            source: 変換元パス（結果記録用）
            bytes_before: 変換前のファイルサイズ

        Returns:
            変換結果
        """
        dest.parent.mkdir(parents=True, exist_ok=True)

        has_alpha = image.mode in ("RGBA", "LA", "PA")

        if has_alpha and self._lossless_alpha:
            # アルファチャンネル付きでロスレスアルファ
            image.save(dest, "WEBP", quality=self._quality, lossless=True)
        elif has_alpha:
            # アルファチャンネル付きだがロスレスでない
            image.save(dest, "WEBP", quality=self._quality)
        else:
            # アルファチャンネルなし→RGB変換
            if image.mode != "RGB":
                image = image.convert("RGB")
            image.save(dest, "WEBP", quality=self._quality)

        bytes_after = self._get_file_size(dest)

        return ConversionResult(
            source_path=source,
            dest_path=dest,
            status=ConversionStatus.SUCCESS,
            bytes_before=bytes_before,
            bytes_after=bytes_after,
        )

    def _save_as_png(
        self,
        image: Image.Image,
        dest: Path,
        source: Path,
        bytes_before: int,
    ) -> ConversionResult:
        """画像をPNG形式で保存する内部メソッド

        PNG形式はロスレス圧縮のため、品質設定は使用されない。

        Args:
            image: 保存するPIL.Imageオブジェクト
            dest: 保存先パス
            source: 変換元パス（結果記録用）
            bytes_before: 変換前のファイルサイズ

        Returns:
            変換結果
        """
        dest.parent.mkdir(parents=True, exist_ok=True)

        # PNG形式で保存（ロスレス）
        # アルファチャンネルがある場合はRGBA、ない場合はRGB
        if image.mode not in ("RGB", "RGBA", "L", "LA", "P", "PA"):
            # 他のモードはRGB/RGBAに変換
            if "A" in image.mode or image.mode == "PA":
                image = image.convert("RGBA")
            else:
                image = image.convert("RGB")

        image.save(dest, "PNG")

        bytes_after = self._get_file_size(dest)

        return ConversionResult(
            source_path=source,
            dest_path=dest,
            status=ConversionStatus.SUCCESS,
            bytes_before=bytes_before,
            bytes_after=bytes_after,
        )

File: internal/converter/image.go
package converter

import (
	"bytes"
	"errors"
	"fmt"
	"image"
	"image/jpeg"
	"image/png"
	"io"
	"os"
	"path/filepath"
	"strings"

	"golang.org/x/image/bmp"

	"github.com/na2na-p/mnemonic/internal/converter/tlg"
)

// ErrTLGDecodeNotImplemented はTLG画像のデコードが未実装であることを示す
// センチネルエラー。
//
// why not: TLG6の本体デコード実装はスコープ外であり、ヘッダのマジックバイト
// 判定とヘッダ解析のみ実装する。TLG5は実装済みのため、このエラーはTLG6形式の
// ファイルに対してのみ返る。
var ErrTLGDecodeNotImplemented = errors.New("TLGデコードは未実装です")

// ErrUnsupportedImageFormat はstdlib/x-imageで対応していない画像拡張子を
// 指定した場合のエラー。
var ErrUnsupportedImageFormat = errors.New("サポートされていない画像形式です")

// ErrTLGInvalidFormat はTLG5/TLG6/SDSのいずれのマジックバイトにも一致しない
// データに対するエラー。
var ErrTLGInvalidFormat = errors.New("TLG形式ではありません")

var (
	tlg5Magic = tlg.TLG5Magic
	tlg6Magic = tlg.TLG6Magic
	sdsMagic  = []byte("TLG0.0\x00sds\x1a")
)

// TLGVersion はTLG画像のバージョンを表す。
type TLGVersion string

// TLGVersionの各値。
const (
	TLGVersionTLG5    TLGVersion = "TLG5"
	TLGVersionTLG6    TLGVersion = "TLG6"
	TLGVersionUnknown TLGVersion = "UNKNOWN"
)

// TLGInfo はTLG画像のメタ情報を表す不変値。
type TLGInfo struct {
	Version  TLGVersion
	Width    int
	Height   int
	HasAlpha bool
}

// TLGImageDecoder はTLG形式の画像ファイルを読み込み、image.Imageへ変換する。
// TLG5およびTLG6形式に対応（TLG6は本体デコード未実装）。SDSコンテナ形式も
// サポートする。
type TLGImageDecoder struct {
	tlg5Decoder *tlg.TLG5Decoder
	tlg6Decoder *tlg.TLG6Decoder
}

// NewTLGImageDecoder はTLGImageDecoderを初期化する。
func NewTLGImageDecoder() *TLGImageDecoder {
	return &TLGImageDecoder{
		tlg5Decoder: tlg.NewTLG5Decoder(),
		tlg6Decoder: tlg.NewTLG6Decoder(),
	}
}

// IsTLGFile はfilePathがTLG5/TLG6/SDS形式かどうかをマジックバイトで判定する。
func (d *TLGImageDecoder) IsTLGFile(filePath string) bool {
	f, err := os.Open(filePath) //nolint:gosec // 呼び出し側が指定したアセットパスを読む用途のため妥当
	if err != nil {
		return false
	}
	defer func() { _ = f.Close() }()

	header := make([]byte, len(tlg5Magic))
	if _, err := io.ReadFull(f, header); err != nil {
		return false
	}

	return bytes.Equal(header, tlg5Magic) || bytes.Equal(header, tlg6Magic) || bytes.Equal(header, sdsMagic)
}

// unwrapSDS はSDSコンテナから内部のTLGデータを抽出する。
//
// SDSコンテナ構造: マジック(11バイト) + チャンクサイズ(4バイト、リトル
// エンディアン) + 内部TLGデータ(TLG5またはTLG6)。
func unwrapSDS(data []byte) []byte {
	if !bytes.HasPrefix(data, sdsMagic) {
		return data
	}

	const sdsHeaderSize = 15 // マジック(11) + チャンクサイズ(4)
	if len(data) < sdsHeaderSize {
		return data
	}

	return data[sdsHeaderSize:]
}

// detectVersion はTLGデータのバージョンを判別する。
func detectVersion(data []byte) TLGVersion {
	switch {
	case bytes.HasPrefix(data, tlg5Magic):
		return TLGVersionTLG5
	case bytes.HasPrefix(data, tlg6Magic):
		return TLGVersionTLG6
	default:
		return TLGVersionUnknown
	}
}

// readTLGSource はfilePathを読み込み、SDSコンテナを解いた生データを返す。
// ファイルが存在しない場合はErrSourceNotFoundを返す。
func readTLGSource(filePath string) ([]byte, error) {
	data, err := os.ReadFile(filePath) //nolint:gosec // 呼び出し側が指定したアセットパスを読む用途のため妥当
	if err != nil {
		return nil, fmt.Errorf("%w: %s", ErrSourceNotFound, filePath)
	}

	return unwrapSDS(data), nil
}

// GetInfo はTLG画像のメタ情報を取得する。
//
// why not: TLG6のヘッダ解析は実装済みのTLG6Decoder.ParseHeaderで完結するため、
// GetInfoはTLG5/TLG6のいずれでもErrTLGDecodeNotImplementedを返さない
// （decode()本体のみが未実装であるため）。
func (d *TLGImageDecoder) GetInfo(filePath string) (TLGInfo, error) {
	data, err := readTLGSource(filePath)
	if err != nil {
		return TLGInfo{}, err
	}

	switch detectVersion(data) {
	case TLGVersionTLG5:
		header, parseErr := d.tlg5Decoder.ParseHeader(data)
		if parseErr != nil {
			return TLGInfo{}, parseErr
		}

		return TLGInfo{Version: TLGVersionTLG5, Width: header.Width, Height: header.Height, HasAlpha: header.Colors == 4}, nil
	case TLGVersionTLG6:
		header, parseErr := d.tlg6Decoder.ParseHeader(data)
		if parseErr != nil {
			return TLGInfo{}, parseErr
		}

		return TLGInfo{Version: TLGVersionTLG6, Width: header.Width, Height: header.Height, HasAlpha: header.Colors == 4}, nil
	default:
		return TLGInfo{}, fmt.Errorf("%w: %s", ErrTLGInvalidFormat, filePath)
	}
}

// Decode はTLG画像をデコードしてimage.Imageを返す。
// TLG6形式の場合はErrTLGDecodeNotImplementedを返す（本体デコード未実装）。
func (d *TLGImageDecoder) Decode(filePath string) (image.Image, error) {
	data, err := readTLGSource(filePath)
	if err != nil {
		return nil, err
	}

	switch detectVersion(data) {
	case TLGVersionTLG5:
		img, decErr := d.tlg5Decoder.Decode(data)
		if decErr != nil {
			return nil, decErr
		}

		return img, nil
	case TLGVersionTLG6:
		// why not: tlg6Decoder.Decode()はマジックバイトが有効な限り常に
		// tlg.ErrTLG6NotImplemented（"TLG6デコードは未実装です"）を返す。
		// これをErrTLGDecodeNotImplementedへさらに%wでラップすると
		// 「TLGデコードは未実装です: TLG6デコードは未実装です」という
		// 冗長な二重メッセージになるため、下位エラーの文言は引き継がず
		// ErrTLGDecodeNotImplementedのみを返す。呼び出し自体は、将来
		// tlg6Decoder.Decodeが実装された際にここを更新し忘れないための
		// フックとして残す。
		_, _ = d.tlg6Decoder.Decode(data)

		return nil, ErrTLGDecodeNotImplemented
	default:
		return nil, fmt.Errorf("%w: %s", ErrTLGInvalidFormat, filePath)
	}
}

// DecodeToFile はTLG画像をデコードしてファイルに保存する。
// destの拡張子が.pngの場合に保存する。
func (d *TLGImageDecoder) DecodeToFile(source, dest string) error {
	img, err := d.Decode(source)
	if err != nil {
		return err
	}

	if err := os.MkdirAll(filepath.Dir(dest), 0o750); err != nil {
		return fmt.Errorf("出力先ディレクトリの作成に失敗しました: %w", err)
	}

	return encodeImageToFile(img, dest)
}

// encodeImageToFile はimgをPNG形式でファイルへ書き出す。
func encodeImageToFile(img image.Image, dest string) error {
	f, err := os.Create(dest) //nolint:gosec // ビルド成果物の出力用途のため妥当
	if err != nil {
		return fmt.Errorf("出力ファイルの作成に失敗しました: %w", err)
	}
	defer func() { _ = f.Close() }()

	ext := strings.ToLower(filepath.Ext(dest))

	switch ext {
	case ".png":
		if err := png.Encode(f, img); err != nil {
			return fmt.Errorf("PNGエンコードに失敗しました: %w", err)
		}
	default:
		return fmt.Errorf("%w: %s", ErrUnsupportedImageFormat, ext)
	}

	return nil
}

// ImageConverter はBMP/JPG/PNG/TLG形式の画像をPNG形式に変換するConverter。
type ImageConverter struct {
	tlgDecoder *TLGImageDecoder
}

// NewImageConverter はImageConverterを初期化する。
//
// why not: krkrsdl2はWebPを読み込めないため、出力形式はPNGに固定する。
func NewImageConverter() *ImageConverter {
	return &ImageConverter{tlgDecoder: NewTLGImageDecoder()}
}

// SupportedExtensions は対応する拡張子の一覧を返す。
//
// JPEG/PNG/BMPはkrkrsdl2でネイティブサポートのため変換対象外
// （feat/exe-icon-extraction 680b27fより。誤ってPNGに変換されたJPEGを
// krkrsdl2のTVPLoadJPEGへ渡すとSIGSEGVでクラッシュする不具合の修正）。
func (c *ImageConverter) SupportedExtensions() []string {
	return []string{".tlg"}
}

// GetOutputExtension は常に.pngを返す。
func (c *ImageConverter) GetOutputExtension(_ string) string {
	return ".png"
}

// CanConvert はfilePathが変換可能かを拡張子で判定する。
func (c *ImageConverter) CanConvert(filePath string) bool {
	ext := strings.ToLower(filepath.Ext(filePath))

	return containsString(c.SupportedExtensions(), ext)
}

// Convert は画像ファイルを指定された形式に変換し、destへ出力する。
//
// why not: 他のConverterと異なり、validateSource・TLG未実装エラー・画像
// デコードの失敗を自身で捕捉せず、呼び出し元(ConversionManager)へerrとして
// 伝播させる。これらの失敗はConversionResultではなくerrとして返す。
//
// why not(decodeSourceの対応拡張子): CanConvert/SupportedExtensionsは.tlg
// のみだが、Convert()自体は.bmp/.jpg/.jpeg/.png/.tlgを直接処理できる
// （ConversionManager経由では.tlg以外はルーティングされないが、
// ConvertFromImage等、呼び出し元がConvertを直接呼ぶ経路のためにdecodeSource
// の分岐は維持する）。
func (c *ImageConverter) Convert(source, dest string) (ConversionResult, error) {
	if err := validateSource(source); err != nil {
		return ConversionResult{}, permanentError(err)
	}

	bytesBefore := getFileSize(source)

	img, err := c.decodeSource(source)
	if err != nil {
		return ConversionResult{}, err
	}

	return c.saveAsPNG(img, dest, source, bytesBefore)
}

// ConvertFromImage はメモリ上のimage.ImageをPNG形式で保存する。
// TLGデコード後の画像変換等、既にデコード済みの画像を直接保存する用途。
func (c *ImageConverter) ConvertFromImage(img image.Image, dest string) (ConversionResult, error) {
	return c.saveAsPNG(img, dest, dest, 0)
}

func (c *ImageConverter) decodeSource(source string) (image.Image, error) {
	ext := strings.ToLower(filepath.Ext(source))
	if ext == ".tlg" {
		img, err := c.tlgDecoder.Decode(source)
		if err != nil {
			// why not: readTLGSourceは読み込み失敗をOSのエラーを捨てて一律ErrSourceNotFoundに
			// するため、ここでは原因（権限不足・削除競合・I/Oエラー）を区別できない。
			// validateSourceで存在は確認済みであり、原因が判別できない以上は他のconverterの
			// 読み込み失敗と同じく再試行側に寄せる。OSのエラーを包むようになった時点で
			// fs.ErrPermission等を恒久扱いに分類し直す。
			if errors.Is(err, ErrSourceNotFound) {
				return nil, err
			}

			return nil, permanentError(err)
		}

		return img, nil
	}

	f, err := os.Open(source) //nolint:gosec // validateSourceで存在確認済みのアセットパスを読む用途のため妥当
	if err != nil {
		return nil, fmt.Errorf("画像ファイルのオープンに失敗しました: %w", err)
	}
	defer func() { _ = f.Close() }()

	switch ext {
	case ".bmp":
		img, err := bmp.Decode(f)
		if err != nil {
			return nil, permanentError(fmt.Errorf("BMP画像のデコードに失敗しました: %w", err))
		}

		return img, nil
	case ".jpg", ".jpeg":
		img, err := jpeg.Decode(f)
		if err != nil {
			return nil, permanentError(fmt.Errorf("JPEG画像のデコードに失敗しました: %w", err))
		}

		return img, nil
	case ".png":
		img, err := png.Decode(f)
		if err != nil {
			return nil, permanentError(fmt.Errorf("PNG画像のデコードに失敗しました: %w", err))
		}

		return img, nil
	default:
		return nil, permanentError(fmt.Errorf("%w: %s", ErrUnsupportedImageFormat, ext))
	}
}

// saveAsPNG は画像をPNG形式で保存する内部メソッド。
//
// why not: Goのimage/png.Encodeは任意のimage.Imageを受け付け、そのカラー
// モデルに応じて適切なPNGを書き出すため、モード変換分岐は不要
// （image/pngが吸収する）。
func (c *ImageConverter) saveAsPNG(img image.Image, dest, source string, bytesBefore int64) (ConversionResult, error) {
	if err := os.MkdirAll(filepath.Dir(dest), 0o750); err != nil {
		return ConversionResult{}, fmt.Errorf("出力先ディレクトリの作成に失敗しました: %w", err)
	}

	f, err := os.Create(dest) //nolint:gosec // ビルド成果物の出力用途のため妥当
	if err != nil {
		return ConversionResult{}, fmt.Errorf("出力ファイルの作成に失敗しました: %w", err)
	}
	defer func() { _ = f.Close() }()

	if err := png.Encode(f, img); err != nil {
		return ConversionResult{}, fmt.Errorf("PNGエンコードに失敗しました: %w", err)
	}

	return ConversionResult{
		SourcePath:  source,
		DestPath:    dest,
		Status:      StatusSuccess,
		BytesBefore: bytesBefore,
		BytesAfter:  getFileSize(dest),
	}, nil
}
